src/persist/src/file.rs

```diff
@@ -182,6 +183,179 @@ impl Persister for FilePersister {
     }
 }
 
+/// A naive implementation of [Buffer] backed by files.
+pub struct FileBuffer {
+    // Store data and sequence number metadata in two separate files
+    dataz: Arc<Mutex<(File, File)>>,
+    seqno: Range<SeqNo>,
+    buf: Vec<u8>,
+}
+
+impl FileBuffer {
+    const BUFFER_PATH: &'static str = "BUFFER";
+    const LOCKFILE_PATH: &'static str = "LOCK";
+    const METADATA_PATH: &'static str = "META";
+
+    /// Returns a new [FileBuffer] which stores files under the given dir.
+    ///
+    /// To ensure directory-wide mutual exclusion, a LOCK file is placed in
+    /// base_dir at construction time. If this file already exists (indicating
+    /// that another FileBuffer is already using the dir), an error is returned
+    /// from `new`.
+    ///
+    /// The contents are `lock_info` are stored in the LOCK file and should
+    /// include anything that would help debug an unexpected LOCK file, such as
+    /// version, ip, worker number, etc.
+    pub fn new<P: AsRef<Path>>(base_dir: P, lock_info: &[u8]) -> Result<Self, Error> {
+        let base_dir = base_dir.as_ref();
+        fs::create_dir_all(&base_dir)?;
+        {
+            let lockfile_path = Self::lockfile_path(&base_dir);
+            let mut lockfile = OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(lockfile_path)?;
+            lockfile.write_all(&lock_info)?;
+            lockfile.sync_all()?;
+        }
+        let buffer_path = Self::buffer_path(&base_dir);
+        let buffer_file = OpenOptions::new()
+            .append(true)
+            .read(true)
+            .create(true)
+            .open(&buffer_path)?;
+
+        let metadata_path = Self::metadata_path(&base_dir);
+        let mut metadata_file = OpenOptions::new()
+            .write(true)
+            .read(true)
+            .create(true)
+            .open(&metadata_path)?;
+
+        // Retrieve the sequence number metadata from the file
+        let mut bytes = Vec::new();
+        metadata_file.read_to_end(&mut bytes)?;
+        let seqno = if bytes.len() != 0 {
+            assert!(bytes.len() == 16);
+            let mut bytes = &bytes[..];
+            let mut buf = [0u8; 8];
+            // Decode sequence number
+            bytes.read_exact(&mut buf)?;
+            let seqno_start = SeqNo(u64::from_le_bytes(buf));
+            bytes.read_exact(&mut buf)?;
+            let seqno_end = SeqNo(u64::from_le_bytes(buf));
+            seqno_start..seqno_end
+        } else {
+            let bytes = [0u8; 16];
+            metadata_file.write_all(&bytes)?;
```

Comment:
need to sync here. also hardcoding 16 bytes of zero is kind of cringe -- maybe its better to extract this out into a `write_seqno` function?